File: IcarianCS/src/Scribe.cs
// Icarian Engine - C# Game Engine
// 
// License at end of file.

using IcarianEngine.Mod;
using IcarianEngine.Rendering.UI;
using System.Collections.Concurrent;
using System.IO;
using System.Xml;

namespace IcarianEngine
{
    public static class Scribe
    {
        static ConcurrentDictionary<string, Font> s_fonts;
        static ConcurrentDictionary<string, string> s_strings;

        static string s_curLanguage;

        /// <summary>
        /// The currently selected language for localization
        /// </summary>
        public static string CurrentLanguage
        {
            get
            {
                return s_curLanguage;
            }
        }

        /// <summary>
        /// If a string exists for the current locale
        /// </summary>
        /// <param name="a_key">The string to find</param>
        /// <returns>If the string exists</returns>
        public static bool StringKeyExists(string a_key)
        {
            return s_strings.ContainsKey(a_key);
        }
        /// <summary>
        /// If a <see cref="IcarianEngine.Rendering.UI.Font" /> exists for the string in the current locale
        /// </summary>
        /// <param name="a_key">The <see cref="IcarianEngine.Rendering.UI.Font" /> to find</param>
        /// <returns>If the <see cref="IcarianEngine.Rendering.UI.Font" /> exists</returns>
        public static bool FontKeyExists(string a_key)
        {
            return s_fonts.ContainsKey(a_key);
        }

        static void LoadFile(string a_path)
        {
            XmlDocument doc = new XmlDocument();
            doc.Load(a_path);

            if (doc.DocumentElement is XmlElement root)
            {
                string language = "default";
                string fontpath = null;
                foreach (XmlAttribute att in root.Attributes)
                {
                    switch (att.Name.ToLower())
                    {
                    case "language":
                    {
                        language = att.Value.ToLower();

                        break;
                    }
                    case "font":
                    {
                        fontpath = att.Value;

                        break;
                    }
                    default:
                    {
                        Logger.IcarianWarning($"Invalid attribute in scribe file: {att.Name} at {a_path}");

                        break;
                    }
                    }
                }
                
                Font font = null;
                if (!string.IsNullOrWhiteSpace(fontpath))
                {
                    font = AssetLibrary.LoadFont(fontpath);
                }

                if (language == "default")
                {
                    foreach (XmlNode node in root.ChildNodes)
                    {
                        XmlElement element = node as XmlElement;
                        if (element == null)
                        {
                            continue;
                        }

                        string name = element.Name;
                        if (!string.IsNullOrWhiteSpace(name))
                        {
                            if (!StringKeyExists(name))
                            {
                                string text = element.InnerText;

                                SetString(name, text);

                                if (font != null)
                                {
                                    SetFont(name, font);
                                }
                            }
                        }
                        else
                        {
                            Logger.IcarianWarning($"Invalid key name in Scribe file at {a_path}");
                        }
                    }
                }
                else if (language == s_curLanguage.ToLower())
                {
                    foreach (XmlNode node in root.ChildNodes)
                    {
                        XmlElement element = node as XmlElement;
                        if (element == null)
                        {
                            continue;
                        }

                        string name = element.Name;
                        if (!string.IsNullOrWhiteSpace(name))
                        {
                            string text = element.InnerText;
                            if (text == null)
                            {
                                text = string.Empty;
                            }

                            SetString(name, text);

                            if (font != null)
                            {
                                SetFont(name, font);
                            }
                        }
                        else
                        {
                            Logger.IcarianWarning($"Invalid key name in Scribe file at {a_path}");
                        }
                    }
                }
            }
        }

        internal static void LoadDirectory(string a_path)
        {
            if (!Directory.Exists(a_path))
            {
                return;
            }

            string[] files = Directory.GetFiles(a_path);
            foreach (string f in files)
            {
                if (Path.GetExtension(f) == ".scrb")
                {
                    LoadFile(f);
                }
            }

            string[] paths = Directory.GetDirectories(a_path);
            foreach (string p in paths)
            {
                LoadDirectory(p);
            }
        }

        /// <summary>
        /// Loads the locale for the language
        /// </summary>
        /// <param name="a_langauge">The language to set the locale to</param>
        public static void SetLanguage(string a_language)
        {
            s_curLanguage = a_language;

            s_fonts = new ConcurrentDictionary<string, Font>();
            s_strings = new ConcurrentDictionary<string, string>();

            LoadDirectory(Path.Combine(ModControl.CoreAssembly.AssemblyInfo.Path, "Scribe"));

            foreach (IcarianAssembly a in ModControl.Assemblies)
            {
                LoadDirectory(Path.Combine(a.AssemblyInfo.Path, "Scribe"));
            }
        }

        /// <summary>
        /// Sets a string
        /// </summary>
        /// <param name="a_key">The string to use as a key</param>
        /// <param name="a_value">The string to use as a value</param>
        public static void SetString(string a_key, string a_value)
        {
            if (StringKeyExists(a_key))
            {
                string s = s_strings[a_key];

                s_strings.TryUpdate(a_key, a_value, s);
            }
            else
            {
                s_strings.TryAdd(a_key, a_value);
            }
        }
        /// <summary>
        /// Gets a string from the locale
        /// </summary>
        /// <param name="a_key">The string to get from the locale</param>
        /// <returns>The locale string. The key on failure</returns>
        public static string GetString(string a_key)
        {
            if (StringKeyExists(a_key))
            {
                return s_strings[a_key];
            }

            return a_key;
        }
        /// <summary>
        /// Gets a formated string from the locale
        /// </summary>
        /// <param name="a_key">The string to get from the locale</param>
        /// <param name="a_args">The arguments to replace the placeholder values with</param>
        /// <returns>The formated string. The key on failure</returns>
        public static string GetStringFormated(string a_key, string[] a_args)
        {
            // Not sure if I should parse the arguments through Scribe or that should be the callers resposibility
            // I am gonna say no as the caller may not always want to do that but I may revist later

            return string.Format(GetString(a_key), a_args);
        }

        /// <summary>
        /// Sets the <see cref="IcarianEngine.Rendering.UI.Font" /> for a string key
        /// </summary>
        /// <param name="a_key">The key</param>
        /// <param name="a_font">The <see cref="IcarianEngine.Rendering.UI.Font" /> for the key</param>
        public static void SetFont(string a_key, Font a_font)
        {
            if (FontKeyExists(a_key))
            {
                Font f = s_fonts[a_key];

                s_fonts.TryUpdate(a_key, a_font, f);
            }
            else
            {
                s_fonts.TryAdd(a_key, a_font);
            }
        }
        /// <summary>
        /// Gets a <see cref="IcarianEngine.Rendering.UI.Font" /> from the locale
        /// </summary>
        /// <param name="a_key">The <see cref="IcarianEngine.Rendering.UI.Font" /> key to get from the locale</param>
        /// <returns>The locale font. Null on failure</returns>
        public static Font GetFont(string a_key)
        {
            if (FontKeyExists(a_key))
            {
                return s_fonts[a_key];
            }

            return null;
        }
    }
}

// MIT License
// 
// Copyright (c) 2024 River Govers
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

File: IcarianNative/src/Scribe.cpp
// Icarian Engine - C# Game Engine
// 
// License at end of file.

#include "Scribe.h"

#include <codecvt>
#include <locale>

#include "Core/IcarianDefer.h"
#include "Runtime/RuntimeManager.h"

Scribe* Scribe::Instance = nullptr;

static std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> converter;

RUNTIME_FUNCTION(void, Scribe, SetInternalLanguage, 
{
    char* language = mono_string_to_utf8(a_language);
    IDEFER(mono_free(language));

    Scribe::SetCurrentLanguage(language);
}, MonoString* a_language)
RUNTIME_FUNCTION(MonoString*, Scribe, GetInternalLanguage,
{
    return mono_string_new_wrapper(Scribe::GetCurrentLanguage().c_str());
})

RUNTIME_FUNCTION(uint32_t, Scribe, GetFontAddr, 
{
    char* key = mono_string_to_utf8(a_key);
    IDEFER(mono_free(key));

    return Scribe::GetFont(key);
}, MonoString* a_key);
RUNTIME_FUNCTION(MonoString*, Scribe, GetString,
{
    char* key = mono_string_to_utf8(a_key);
    IDEFER(mono_free(key));

    const std::u32string str = Scribe::GetString(key);
    if (!str.empty())
    {
        return mono_string_from_utf32((mono_unichar4*)str.c_str());
    }

    return NULL;
}, MonoString* a_key)
RUNTIME_FUNCTION(MonoString*, Scribe, GetStringFormated, 
{
    const uintptr_t size = mono_array_length(a_args);

    char32_t** args = nullptr;
    if (size > 0)
    {
        args = new char32_t*[size];
        for (uintptr_t i = 0; i < size; ++i)
        {
            args[i] = (char32_t*)mono_string_to_utf32(mono_array_get(a_args, MonoString*, i));
        }
    }

    IDEFER(
    if (args != nullptr)
    {
        for (uint32_t i = 0; i < size; ++i)
        {
            mono_free(args[i]);
        }

        delete[] args;
    });

    char* key = mono_string_to_utf8(a_key);
    IDEFER(mono_free(key));

    const std::u32string str = Scribe::GetStringFormated(key, args, (uint32_t)size);

    if (!str.empty())
    {
        return mono_string_from_utf32((mono_unichar4*)str.c_str());
    }

    return NULL;
}, MonoString* a_key, MonoArray* a_args)

RUNTIME_FUNCTION(void, Scribe, SetFont,
{
    char* key = mono_string_to_utf8(a_key);
    IDEFER(mono_free(key));

    Scribe::SetFont(key, a_value);
}, MonoString* a_key, uint32_t a_value)
RUNTIME_FUNCTION(void, Scribe, SetString,
{
    char* key = mono_string_to_utf8(a_key);
    IDEFER(mono_free(key));
    char32_t* value = (char32_t*)mono_string_to_utf32(a_value);
    IDEFER(mono_free(value));

    Scribe::SetString(key, value);
}, MonoString* a_key, MonoString* a_value)

RUNTIME_FUNCTION(uint32_t, Scribe, Exists,
{
    char* key = mono_string_to_utf8(a_key);
    IDEFER(mono_free(key));

    return (uint32_t)Scribe::KeyExists(key);
}, MonoString* a_key)

Scribe::Scribe()
{
    
}
Scribe::~Scribe()
{

}

void Scribe::Init()
{
    if (Instance == nullptr)
    {
        Instance = new Scribe();

        BIND_FUNCTION(IcarianEngine, Scribe, SetInternalLanguage);
        BIND_FUNCTION(IcarianEngine, Scribe, GetInternalLanguage);

        BIND_FUNCTION(IcarianEngine, Scribe, GetFontAddr);
        BIND_FUNCTION(IcarianEngine, Scribe, GetString);
        BIND_FUNCTION(IcarianEngine, Scribe, GetStringFormated);

        BIND_FUNCTION(IcarianEngine, Scribe, SetFont);
        BIND_FUNCTION(IcarianEngine, Scribe, SetString);

        BIND_FUNCTION(IcarianEngine, Scribe, Exists);
    }
}
void Scribe::Destroy()
{
    if (Instance != nullptr)
    {
        delete Instance;
        Instance = nullptr;
    }
}

std::string Scribe::GetCurrentLanguage()
{
    IVERIFY(Instance != nullptr);

    const SharedThreadGuard g = SharedThreadGuard(Instance->m_lock);

    return Instance->m_curLang;
}
void Scribe::SetCurrentLanguage(const std::string_view& a_language)
{
    IVERIFY(Instance != nullptr);

    const ThreadGuard g = ThreadGuard(Instance->m_lock);

    Instance->m_curLang = std::string(a_language);
}

uint32_t Scribe::GetFont(const std::string_view& a_key)
{
    IVERIFY(Instance != nullptr);

    const std::string key = std::string(a_key);
    if (Instance->m_fonts.Exists(key))
    {
        return Instance->m_fonts[key];
    }

    return -1;
}

std::u32string Scribe::GetString(const std::string_view& a_key)
{
    IVERIFY(Instance != nullptr);
    
    const std::string key = std::string(a_key);
    if (Instance->m_strings.Exists(key))
    {
        return Instance->m_strings[key];
    }

    return converter.from_bytes(key);
}
std::u32string Scribe::GetStringFormated(const std::string_view& a_key, const char32_t* const* a_args, uint32_t a_count)
{
    std::u32string str = GetString(a_key);

    if (str.empty())
    {
        return str;
    }

    for (uint32_t i = 0; i < a_count; ++i)
    {
        const std::u32string iStr = converter.from_bytes("{" + std::to_string(i) + "}");
        const size_t size = iStr.size();

        while (true)
        {
            const std::size_t pos = str.find(iStr);
            if (pos == std::u32string::npos)
            {
                break;
            }

            str.replace(pos, size, a_args[i]);
        }   
    }

    return str;
}
std::u32string Scribe::GetStringFormated(const std::string_view& a_key, const std::u32string* a_args, uint32_t a_count)
{
    std::u32string str = GetString(a_key);

    if (str.empty())
    {
        return str;
    }

    for (uint32_t i = 0; i < a_count; ++i)
    {
        const std::u32string iStr = converter.from_bytes("{" + std::to_string(i) + "}");
        const size_t size = iStr.size();

        while (true)
        {
            const std::size_t pos = str.find(iStr);
            if (pos == std::u32string::npos)
            {
                break;
            }

            str.replace(pos, size, a_args[i]);
        }   
    }

    return str;
}

// MIT License
// 
// Copyright (c) 2024 River Govers
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.